src/plumbing/graph.cljc

```diff
@@ -125,28 +125,45 @@
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Compiling and running graphs
 
+#?(:clj (declare interpreted-eager-compile))
 #?(:clj
 (defn eager-compile
   "Compile graph specification g to a corresponding fnk that is optimized for
    speed. Wherever possible, fnks are called positionally, to reduce the
    overhead of creating and destructuring maps, and the return value is a
    record, which is much faster to create and access than a map.  Compilation
    is relatively slow, however, due to internal calls to 'eval'."
-  [g]
-  (if (fn? g)
-    g
-    (let [g (for [[k sub-g] (->graph g)]
-              [k (eager-compile sub-g)])]
-      (graph-positional/positional-flat-compile (->graph g))))))
+  ([g] (eager-compile g {}))
+  ([g {:keys [positional-limit]
+       :or {positional-limit graph-positional/max-graph-size}}]
+   (let [eager-compile (fn eager-compile [g]
+                         (when (some? g)
+                           (if (fn? g)
+                             g
+                             (let [g* (for [[k sub-g] (->graph g)]
+                                        (do (assert sub-g)
+                                            [k (eager-compile sub-g)]))]
+                               (when (every? second g*)
+                                 (let [g (->graph g*)]
+                                   (when (<= (-> g pfnk/output-schema count)
+                                             positional-limit)
+                                     (graph-positional/positional-flat-compile g))))))))]
+     (assert g)
+     (or (eager-compile g)
+         (interpreted-eager-compile g))))))
 
 #?(:clj
 (defn positional-eager-compile
   "Like eager-compile, but produce a non-keyword function that can be called
    with args in the order provided by arg-ks, avoiding the overhead of creating
    and destructuring a top-level map.  This can yield a substantially faster
-   fn for Graphs with very computationally inexpensive node fnks."
+   fn for Graphs with very computationally inexpensive node fnks.
+  
+  Warning: if any level of g exceeds `graph-positional/max-graph-size`, compilation
+  may fail. Do not use for arbitrarily large graphs."
   [g arg-ks]
-  (fnk-impl/positional-fn (eager-compile g) arg-ks)))
+  (assert (<= (count arg-ks) 20) (str "positional-eager-compile can compile up to 20 arguments"))
+  (fnk-impl/positional-fn (eager-compile g {:positional-limit Double/POSITIVE_INFINITY}) arg-ks)))
```

Comment:
Why the infinity here?

Reply:
Based on my tests, `positional-fn` requires a graph compiled positionally. So here we need to force `eager-compile` to never switch to interpreted mode, which doesn't seem to compile positionally. The infinity here the current way this is enforced.

Of course it would be better to switch to a scalable positional compilation, but I don't think we have one available.

Reply:
Added this as a comment.